packages/graphiql-react/src/editor/components/image-preview.tsx

```diff
@@ -0,0 +1,97 @@
+import type { Token } from 'codemirror';
+import { useEffect, useRef, useState } from 'react';
+
+type ImagePreviewProps = { token: Token };
+
+type ImagePreviewState = {
+  width: number | null;
+  height: number | null;
+  src: string | null;
+  mime: string | null;
+};
+
+export function ImagePreview(props: ImagePreviewProps) {
+  const [state, setState] = useState<ImagePreviewState>({
+    width: null,
+    height: null,
+    src: null,
+    mime: null,
+  });
+
+  const ref = useRef<HTMLImageElement>(null);
+
+  function updateMetadata() {
+    if (!ref.current) {
+      return;
+    }
+
+    const width = ref.current.naturalWidth;
+    const height = ref.current.naturalHeight;
+    const src = ref.current.src;
+
+    if (src !== state.src) {
+      setState(current => ({ ...current, src }));
+      fetch(src, { method: 'HEAD' }).then(response => {
+        setState(current => ({
+          ...current,
+          mime: response.headers.get('Content-Type'),
+        }));
+      });
+    }
+
+    if (width !== state.width || height !== state.height) {
+      setState(current => ({ ...current, height, width }));
+    }
+  }
+
+  useEffect(() => {
+    updateMetadata();
+  });
```

Comment:
should this run on every render?

Reply:
Tbh it was like that before and I blindly migrated the logic 😄 But this is a good chance to break that up a bit, we can be more resourceful here 👍

Reply:
Yes splitting it up between the fetch and a resizeObserver might be a good shout, WDYT?

Reply:
no other benefits, wasn't sure whether you needed to respond to the user resizing the window